Enhance troubleshooting for SSH key injection

ldomb opened the pull request:
Added some more documentation around ssh. I don't think its clear to everyone that "core" is the user needed for ssh nor howto generate the config for the sshKey used in the installation.

ldomb commented:
/assign @rajatchopra

wking commented on docs/user/troubleshooting.md:
```diff
@@ -52,7 +52,11 @@ If no pods are shown, etcd will need to be [investigated](#etcd-is-not-running).
 ### Unable to SSH into Master Nodes
 
-In order to SSH into the master nodes, it is necessary to include an administrator's SSH key during the installation. If SSH authentication is failing, ensure that the proper SSH key is being used.
+In order to SSH into the master nodes as user `core`, it is necessary to include an administrator's SSH key during the installation. This key can be defined in the `install-config.yml` under the variable `sshKey`. The install-config.yml can be generated by running 
```
Mentioning `core` is good.  But I don't think install-config edits are the easiest way to inject the SSH key.  Did you skip the wizard prompt and leave [the associated environment variables][1] unset?

[1]: https://github.com/openshift/installer/blame/v0.5.0/docs/user/environment-variables.md#L37-L41

ldomb replied:
I did not, when it asks you for the ssh key during the install 
? SSH Public Key  [Use arrows to move, type to filter, ? for more help]
  /Users/dobby/.ssh/id_rsa.pub

It actually does not add it in the config ( terraform vars) ass SSH keys. 

Also the env var of OPENSHIFT_INSTALL_SSH_PUB_KEY did not work for me as well. Thus I added it to the config. That worked

ldomb replied:
Tested again and realized the ssh-key was my fault. The installer asks you for a key and shows the path "/Users/dobby/.ssh/id_rsa.pub" to the key. It does not use this path. So if you leave it empty its none. So agree the install-config.yml is not the easiest. Will close pull request and rewrite.

ldomb commented:
doc needs to be ajusted

wking commented on docs/user/troubleshooting.md:
```diff
@@ -52,7 +52,9 @@ If no pods are shown, etcd will need to be [investigated](#etcd-is-not-running).
 ### Unable to SSH into Master Nodes
 
-In order to SSH into the master nodes, it is necessary to include an administrator's SSH key during the installation. If SSH authentication is failing, ensure that the proper SSH key is being used.
+In order to SSH into the master nodes as user `core`, it is necessary to include an administrator's SSH key during the installation. When asked by the installation wizzard for the ssh-key, make sure you define the absolute path to your key. This key will be added to the core user's authorized_keys file. If installed in aws you will not see this key attached to your instance as a keypair in the aw ...
```
nits: "ssh" -> "SSH".  "SSH" is the protocol; `ssh` is just the executable.  And "core" -> "`core`" and "authorized_keys" -> "`authorized_keys`".  And "aws" -> "AWS".

Re: "absolute path", [the wizard prompt is a `Select`][1] of your `~/.ssh/*.pub`.  I don't think it allows free-form input (or at least, it didn't allow it in my tests).

[1]: https://github.com/openshift/installer/blob/b4f5ceb6bfde8d3dc0e29f708e0494488ea37ee0/pkg/asset/installconfig/ssh.go#L99

kikisdeliveryservice replied:
minor: "wizzard" -> "wizard"

kikisdeliveryservice commented on docs/user/troubleshooting.md:
```diff
@@ -52,7 +52,9 @@ If no pods are shown, etcd will need to be [investigated](#etcd-is-not-running).
-In order to SSH into the master nodes, it is necessary to include an administrator's SSH key during the installation. If SSH authentication is failing, ensure that the proper SSH key is being used.
+In order to SSH into the master nodes as user `core`, it is necessary to include an administrator's SSH key during the installation. When asked by the installation wizzard for the ssh-key, make sure you define the absolute path to your key. This key will be added to the core user's authorized_keys file. If installed in aws you will not see this key attached to your instance as a keypair in the aw ...
+
+If SSH authentication is failing, ensure that the proper SSH key is being used.
```
Could we add something like: "Verify your SSH Key by using `oc get configmap -n kube-system cluster-config-v1 -o yaml` etc...? Because I don't think it's clear for people how to verify the key without restarting the install.

kikisdeliveryservice replied:
I think we should still add something like this to the documentation.

wking replied:
With the `oc get configmap ...` and associated lead-in above, I think we can drop this line.

wking commented:
@ldomb, do you have time to come back to this and address the remaining feedback?  Just had another report [here][1] from someone who could have used the "user is `core`" documentation.  If you don't have time, do you mind if I carry your commit into a new PR that addresses the feedback?

[1]: https://groups.google.com/d/msg/openshift-4-dev-preview/5wgpUsoxzDw/2PR5hBYiCAAJ

wking commented on docs/user/troubleshooting.md:
```diff
@@ -50,9 +50,9 @@ sudo crictl pods --name=etcd-member
 If no pods are shown, etcd will need to be [investigated](#etcd-is-not-running).
 
-### Unable to SSH into Master Nodes
+### Unable to access the Master Nodes via SSH 
```
nit: trailing whitespace (and also in the next paragraph).

I'd rather leave this header alone (since I'm fine with "SSH" as a verb).  But if you want to update it, you'll also need to update references to it:

```console
$ git grep unable-to origin/pr/795
origin/pr/795:docs/user/troubleshooting.md:If SSH access to the master nodes isn't available, that will need to be [investigated next](#unable-to-ssh-into-master-node).
```

And it looks like that link is already broken in master because it uses `-node` instead of `-nodes`.  I've filed #980 to fix that, so you don't have to worry about it if you leave the header alone.

wking commented on docs/user/troubleshooting.md:
```diff
@@ -50,9 +50,9 @@ sudo crictl pods --name=etcd-member
+### Unable to access the Master Nodes via SSH 
 
-In order to SSH into the master nodes as user `core`, it is necessary to include an administrator's SSH key during the installation. When asked by the installation wizzard for the ssh-key, make sure you define the absolute path to your key. This key will be added to the core user's authorized_keys file. If installed in aws you will not see this key attached to your instance as a keypair in the aw ...
+In order to ssh into the master nodes as user **core**, it is necessary to include an administrator's ssh-key during the installation. When asked by the installation wizard for the ssh-key, make sure you select your ssh-key from the wizard prompt **~/.ssh/*.pub**. This key will be added to the **core** user's **authorized_keys** file. If installed in AWS you will **not** see this key attached to  ...
```
This is moving in the wrong direction with `SSH` -> `ssh`.  Per [this comment][1], you should be using "SSH" (for the protocol) everywhere unless you are writing `ssh ...` for a shell command.

Also, `core` is more appropriate than **core**, because the username is literal [code][2] you could paste into shell scripts.  Similarly for your paths (`~/.ssh/*.pub` and `authorized_keys`).

*not* would be more appropriate than **not**, because this is for local [emphasis][3].  For example, see "[However, it should *not* be used...][4]" and "[This is *not* a drill][5]".  And [the AWS docs][6] use "key pair", not "keypair".  So I'd rather see something like:

> The public key is placed in `authorized_keys` by Ignition; it is not configured via platform-specific approaches like [AWS key pairs][6].

instead of your last two sentences.

[1]: https://github.com/openshift/installer/pull/795#discussion_r239547838
[2]: https://daringfireball.net/projects/markdown/syntax#code
[3]: https://daringfireball.net/projects/markdown/syntax#em
[4]: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/strong#Usage_notes
[5]: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/em
[6]: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-key-pairs.html

kikisdeliveryservice commented:
Is there a reason why you didn't add my suggestion here?: https://github.com/openshift/installer/pull/795#discussion_r239914737

ldomb commented:
Added section for comment 795

ldomb commented:
No no reason, added it to the doc

On Wed, Jan 2, 2019 at 3:07 PM Kirsten <notifications@github.com> wrote:

> Is there a reason why you didn't add my suggestion here?: #795 (comment)
> <https://github.com/openshift/installer/pull/795#discussion_r239914737>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/installer/pull/795#issuecomment-450970374>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABarZiIW3W8E3mXVWZtTGBdrc4kkhGTDks5u_RGcgaJpZM4ZFJDW>
> .
>

wking commented on docs/user/troubleshooting.md:
```diff
@@ -50,9 +50,15 @@ sudo crictl pods --name=etcd-member
+### Unable to access the Master Nodes via SSH 
 
-In order to SSH into the master nodes, it is necessary to include an administrator's SSH key during the installation. If SSH authentication is failing, ensure that the proper SSH key is being used.
+In order to SSH into the master nodes as user `core`, it is necessary to include an administrator's ssh-key during the installation. When asked by the installation wizard for the ssh-key, make sure you select your ssh-key from the wizard prompt `~/.ssh/*.pub`. This key will be added to the `core` user's `authorized_keys` file. The public key is placed in authorized_keys by Ignition and is not con ...
```
"thefollowing" -> "the following".  Or just end with:

> You can verify the configured SSH key with:
>
> ```sh
> oc get configmap ...
> ```

There's also an `authorized_keys` that still needs backticks.

I'd prefer to use reference-style links for external links (like [we do now for `kubernetes-debug`][1]), but if you're tired of tweaking, I'm ok with the inline link.

[1]: https://github.com/openshift/installer/blame/f0066c4ae126346d986dfbb26f7282bea2819f82/docs/user/troubleshooting.md#L254

wking commented on docs/user/troubleshooting.md:
````diff
@@ -50,9 +50,15 @@ sudo crictl pods --name=etcd-member
+In order to SSH into the master nodes as user `core`, it is necessary to include an administrator's ssh-key during the installation. When asked by the installation wizard for the ssh-key, make sure you select your ssh-key from the wizard prompt `~/.ssh/*.pub`. This key will be added to the `core` user's `authorized_keys` file. The public key is placed in authorized_keys by Ignition and is not con ...
+
+```sh
+oc get configmap -n kube-system cluster-config-v1 -o yaml | grep ssh-rsa
````
It doesn't have to be an RSA key.  Unfortunately it's hard to get into the the unstructutured YAML string, but you can do:

```sh
oc get configmap -o "jsonpath={.data['install-config']}" -n kube-system cluster-config-v1 | grep -A1 sshKey
```

to be a bit more robust.  I expect at some point we'll move this into a structured config.openshift.io custom resource, #680.

wking commented:
I've pushed 73aa0e4b7a7 with the remaining changes I'd like to see here, in case squashing that in is easier than working through GitHub review comments.  You should be able to add that commit to your branch with:

```console
$ git pull https://github.com/wking/openshift-installer.git ssh-detail-fixups
```

if it looks good to you.  Then squash this branch down to a single commit, and we should be good to go :).

ldomb commented:
Squashed commits and created a new PR

https://github.com/openshift/installer/pull/986

Also added your changes. But had to modify the links as you have two [][]
instead of []()



On Thu, Jan 3, 2019 at 3:01 AM W. Trevor King <notifications@github.com>
wrote:

> I've pushed 73aa0e4
> <https://github.com/openshift/installer/commit/73aa0e4b7a7c8df95a61092e5d8f70c2a96d51bd>
> with the remaining changes I'd like to see here, in case squashing that in
> is easier than working through GitHub review comments. You should be able
> to add that commit to your branch with:
>
> $ git pull https://github.com/wking/openshift-installer.git ssh-detail-fixups
>
> if it looks good to you. Then squash this branch down to a single commit,
> and we should be good to go :).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/installer/pull/795#issuecomment-451075590>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABarZvElQ2OaqWPyazlDF79WeDeKnesZks5u_bjtgaJpZM4ZFJDW>
> .
>

wking commented:
> Squashed commits and created a new PR

In the future, it's better to force-push to your existing branch instead of creating a new pull request.  That way it's easier for folks starting from a landed commit to find the PR discussion.

> But had to modify the links as you have two `[][]` instead of `[]()`

Those were intentional.  See [the docs][1] for *reference-style links* and the *implicit link name shortcut*.  But whatever, inline links work too ;).

[1]: https://daringfireball.net/projects/markdown/syntax#link
